crates/hir_ty/src/lib.rs

```diff
@@ -561,6 +559,34 @@ impl TypeWalk for TraitRef {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TypeOutlives {
+    /// The type which must outlive the given lifetime.
+    pub ty: Ty,
+    /// The lifetime which the type must outlive.
+    pub lifetime: Lifetime,
+}
+
+impl TypeWalk for TypeOutlives {
+    fn walk(&self, f: &mut impl FnMut(&Ty)) {
+        self.ty.walk(f);
+    }
+
+    fn walk_mut_binders(
+        &mut self,
+        f: &mut impl FnMut(&mut Ty, DebruijnIndex),
+        binders: DebruijnIndex,
+    ) {
+        self.ty.walk_mut_binders(f, binders);
+    }
+}
+
+impl TypeWalk for LifetimeOutlives {
+    fn walk(&self, _: &mut impl FnMut(&Ty)) {}
+
+    fn walk_mut_binders(&mut self, _: &mut impl FnMut(&mut Ty, DebruijnIndex), _: DebruijnIndex) {}
+}
```

Comment:
Our `TypeWalk` is basically a very weak version of chalks `Fold` right? Hence I did this implementation so this works with `do_canonicalize`